src/python/pants/backend/jvm/targets/jvm_app.py

```diff
@@ -229,10 +50,6 @@ def __init__(self,
       Class-Path set to those jars.
     :param string archive: Create an archive of this type from the bundle.
     """
-    if archive and archive not in Archive.TYPE_NAMES:
```

Comment:
It looks like the `archive`, `binary`, `basename`, `bundles`, etc arguments should be lifted into the baseclass, as they are/should-be shared between subclasses. `deployjar` is I think a JVM-specific concept though, so can stay.